### PR TITLE
feat: implement `CreateVolume`

### LINKED ISSUE #30: Support CreateVolume
A complete csi plugin needs to finish its controllerserver.go
https://github.com/kubernetes-csi/csi-driver-nfs/blob/master/pkg/nfs/controllerserver.go.

We can use the code in https://github.com/kubernetes-incubator/external-storage/blob/master/nfs-client/cmd/nfs-client-provisioner/provisioner.go as comparison.

### PR DESCRIPTION
**Pull request by mayankshah1607**
Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Implements `CreateVolume`

**Which issue(s) this PR fixes**:

Fixes #30

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add implementation for `CreateVolume` and `DeleteVolume`
```

### CODE REVIEW

**Comment by mayankshah1607** [comment]
@andyzhangx @msau42  I have made small modifications to @msau42 's work on implementing `CreateVolume`- https://github.com/msau42/csi-driver-nfs/commit/da0e3d682ea34fd8cbcdd042d0f88235dde61bb5
 Could you please take a look? If things look fine, I shall open a follow-up PR implementing `DeleteVolume` as well
Thanks

**Comment by mayankshah1607** [comment]
As this PR introduces unit tests, I was wondering if we should enable them on Github Actions. We could move on to prow once we start adding more tests.

**Comment by andyzhangx** [comment]
> As this PR introduces unit tests, I was wondering if we should enable them on Github Actions. We could move on to prow once we start adding more tests.

sure, go ahead

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [inline pkg/nfs/controllerserver.go:70]: remove the logging since it's not necessary

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [inline pkg/nfs/controllerserver.go:101]: make default as 0777

**Review by andyzhangx — COMMENTED** [review]
pls add `ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME` here:
https://github.com/kubernetes-csi/csi-driver-nfs/blob/7fdf12763de819564a19bffdc06c043fac7250a7/pkg/nfs/nfs.go#L72

also have you tried manually to make sure it works?

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [inline pkg/nfs/controllerserver.go:284]: what if `workingMountDir` is empty?

Comment (andyzhangx) [reply pkg/nfs/controllerserver.go:284]: it's better define a default path, e.g. `/mnt`

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [inline pkg/nfs/controllerserver.go:300]: why this `filepath.Join` takes 3 parameters?

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [inline pkg/nfs/controllerserver.go:223]: use `paramServer`

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [inline pkg/nfs/controllerserver.go:224]: use `paramShare`

**Comment by mayankshah1607** [comment]
> pls add ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME here:

@andyzhangx On doing so, the sanity tests are failing with:
```
I1008 09:23:27.523659    4253 utils.go:48] GRPC call: /csi.v1.Controller/CreateVolume
I1008 09:23:27.524697    4253 utils.go:49] GRPC request: {"capacity_range":{"required_bytes":10737418240},"name":"sanity-controller-create-single-with-capacity-0E4531D2-F7FFA39B","volume_capabilities":[{"AccessType":{"Mount":{}},"access_mode":{"mode":1}}]}
E1008 09:23:27.543171    4253 utils.go:52] GRPC error: rpc error: code = InvalidArgument desc = server is a required parameter
```

File: pkg/nfs/controllerserver.go
Comment (msau42) [reply pkg/nfs/controllerserver.go:300]: This is the path of the share that was created and put into the returned VolumeContext. ie, if you have Storageclass param `base-dir = myvol`, `server = 10.10.10.10`, then after CreateVolume finishes, the VolumeContext returned should be: `server = 10.10.10.10`, `share = /myvol/foo-subdir`

**Comment by msau42** [comment]
> As this PR introduces unit tests, I was wondering if we should enable them on Github Actions. We could move on to prow once we start adding more tests.

Unit tests should already be run as part of pull-kubernetes-csi-csi-driver-nfs. You can double check the build log of the job to see what gets tested

**Comment by andyzhangx** [comment]
could you fix the sanity test failures? Thanks.

File: pkg/nfs/controllerserver.go
Comment (mayankshah1607) [inline pkg/nfs/controllerserver.go:269]: @andyzhangx The sanity tests errors are coming here. If we want to keep this validation, we will have to update the existing sanity tests to use `server` and `baseDir` as request parameters. 
Is that ok?

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [inline pkg/nfs/controllerserver.go:49]: use one `const (` struct

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [inline pkg/nfs/controllerserver.go:70]: remove this empty line

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [reply pkg/nfs/controllerserver.go:269]: need to define a testing nfs-server in sanity test, otherwise even we ignore this error, `CreateVolume` would fail finally.

Comment (mayankshah1607) [reply pkg/nfs/controllerserver.go:269]: And what would the values of these parameters be? Would adding these to the [defaults](func NewTestConfig() TestConfig {) break any existing tests?

Comment (mayankshah1607) [reply pkg/nfs/controllerserver.go:267]: > need to set up a testing nfs-server in sanity test, otherwise even we ignore this error, CreateVolume would fail finally.

Can we run the [existing nfs server example](https://github.com/kubernetes-csi/csi-driver-nfs/tree/master/examples/kubernetes/nfs-provisioner) before initiating sanity tests? How would we feed in the `server` and `baseDir` parameters?

Comment (andyzhangx) [reply pkg/nfs/controllerserver.go:267]: I think yes, and I have modified the nfs server example: https://github.com/kubernetes-csi/csi-driver-nfs/pull/61, now could use `nfs-server.default.svc.cluster.local:/` in sanity test.

Need to add a flag in driver, if it's sanity test, use `nfs-server.default.svc.cluster.local:/`

Comment (andyzhangx) [reply pkg/nfs/controllerserver.go:267]: ah, in sanity test, there is no k8s cluster, it has only one machine, so you need to set up a containerized nfs-server, and use `127.0.0.1:/` for nfs-server.

Comment (mayankshah1607) [reply pkg/nfs/controllerserver.go:267]: Seems like there's an issue with running docker containers within the sanity-test container:
```
docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.
```
Is docker in docker the right way to go about provisioning an nfs server for sanity tests?

**Comment by mayankshah1607** [comment]
@andyzhangx The sanity test is failing to start docker:
```
docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.
```

It looks like we're going to have to modify the present prow job to opt-in for Docker in Docker by adding the `preset-dind-enabled: "true` label. Example - https://github.com/kubernetes/test-infra/pull/9983/files
If this approach for running an NFS server is desirable, shall I open a PR against k/test-infra?

**Comment by andyzhangx** [comment]
> @andyzhangx @msau42 The sanity test is failing to start docker:
> 
> ```
> docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.
> ```
> 
> It looks like we're going to have to modify the present prow job to opt-in for Docker in Docker by adding the `preset-dind-enabled: "true"` label. Example - https://github.com/kubernetes/test-infra/pull/9983/files
> If this approach for running an NFS server is desirable, shall I open a PR against k/test-infra?

@mayankshah1607 what about run sanity test in github action, that could be easier to configure?

File: test/sanity/run-test.sh
Comment (andyzhangx) [inline test/sanity/run-test.sh:37]: may need 111 UDP port?

**Comment by andyzhangx** [comment]
great job, only two sanity test failures left:
```
Summarizing 2 Failures:

[Fail] Controller Service [Controller Server] CreateVolume [It] should not fail when requesting to create a volume with already existing name and same capacity. 
/home/runner/go/pkg/mod/github.com/kubernetes-csi/csi-test@v2.2.0+incompatible/pkg/sanity/controller.go:567

[Fail] Controller Service [Controller Server] CreateVolume [It] should fail when requesting to create a volume with already existing name and different capacity. 
/home/runner/go/pkg/mod/github.com/kubernetes-csi/csi-test@v2.2.0+incompatible/pkg/sanity/controller.go:646
```

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [reply pkg/nfs/controllerserver.go:97]: I think we could skip this case since `CreateVolume` could not set the capacity

Comment (andyzhangx) [reply pkg/nfs/controllerserver.go:97]: use example here to skip this case: https://github.com/kubernetes-csi/csi-driver-smb/blob/master/test/sanity/run-test.sh#L47

**Comment by msau42** [comment]
A disadvantage of Github actions is that the k8s merge bot doesn't block on it. Have we tried fixing the prow job config to enable docker in docker and that didn't work? I guess once question is how well running multiple nfs servers in different containers will work.

File: pkg/nfs/controllerserver.go
Comment (msau42) [reply pkg/nfs/controllerserver.go]: Is this a problem with the CSI spec or with the test case? Can we open bugs to track fixing either the spec or test to work for our use case, which is not unusual

Comment (mayankshah1607) [reply pkg/nfs/controllerserver.go]: I suspect that the problem could be due to the way this implementation of `CreateVolume` is performing this particular check (with the `volume` package) w.r.t the test. I'm not sure if my understanding is correct, but if the test isn't really provisioning a volume, should the `NewMetricStatFs` method expect to work correctly?

File: pkg/nfs/controllerserver.go
Comment (mayankshah1607) [inline pkg/nfs/controllerserver.go:95]: @msau42 I was also a little confused about this part. Should the response not contain information about capacity?

Comment (msau42) [reply pkg/nfs/controllerserver.go:95]: this was a workaround until the issue in external-provisioner was fixed. Now that it is fixed, I think we can return 0 capacity.

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [reply pkg/nfs/controllerserver.go]: while such `req >= actual size` implementation would still make sanity test fail, so let's disable this test for this driver

**Comment by mayankshah1607** [comment]
@andyzhangx @msau42 It looks like the sanity tests are successfully passing after enabling DinD in Prow. Should we remove the GH Actions workflow or keep both for now for ensuring stability?

Also, returning `CapacityBytes` as 0 in `CreateVolumeResponse` causes 2 tests to fail:

```
Controller Service [Controller Server] CreateVolume 
  should return appropriate values SingleNodeWriter WithCapacity 1Gi Type:Mount

Controller Service [Controller Server] CreateVolume 
  should fail when requesting to create a volume with already existing name and different capacity.
```
We could either skip these as well, or return the the same value as the request

**Comment by andyzhangx** [comment]
> should return appropriate values SingleNodeWriter WithCapacity 1Gi Type:Mount

let's keep the GH action since it won't bring any more effort.
And about case `should return appropriate values SingleNodeWriter WithCapacity 1Gi Type:Mount`, is it possible to return same `CapacityBytes` as original request?

**Comment by mayankshah1607** [comment]
> And about case should return appropriate values SingleNodeWriter WithCapacity 1Gi Type:Mount, is it possible to return same CapacityBytes as original request?

Yes, that's what I have done right now for the tests to pass. Just wanted to confirm if this would be a desirable behavior of the client, so we can proceed with getting this one merged if everything looks okay

File: .github/workflows/sanity.yaml
Comment (andyzhangx) [inline .github/workflows/sanity.yaml:14]: we don't need this change?

File: pkg/nfs/controllerserver.go
Comment (andyzhangx) [inline pkg/nfs/controllerserver.go:84]: is `validCapacity` still necessary since some sanity tests are not running any more, also the capacity is actually not controlled by this driver.

File: pkg/nfs/utils.go
Comment (andyzhangx) [inline pkg/nfs/utils.go:23]: in current implementation, `workingMountDir` is always `/tmp`, so maybe we don't need this field, just use a const var `workingMountDir` is enough

File: .github/workflows/sanity.yaml
Comment (mayankshah1607) [reply .github/workflows/sanity.yaml:14]: This change was required so that the sanity test prow job could reuse this functionality

File: pkg/nfs/controllerserver.go
Comment (mayankshah1607) [reply pkg/nfs/controllerserver.go:84]: yes, I will remove it then

**Comment by mayankshah1607** [comment]
@andyzhangx I have made these changes. PTAL, thanks!

File: pkg/nfs/utils.go
Comment (msau42) [reply pkg/nfs/utils.go]: I believe it was parameterized this way to facilitate unit tests

File: pkg/nfs/controllerserver.go
Comment (msau42) [reply pkg/nfs/controllerserver.go]: I'm fine with disabling the test to unblock work, but we need to follow up with correcting the spec/test/driver. The test is supposed to make sure drivers are CSI compliant so if it's not working, we need to figure out why.

File: examples/kubernetes/storageclass-nfs.yaml
Comment (msau42) [inline examples/kubernetes/storageclass-nfs.yaml:10]: Delete is supported now right?

Comment (mayankshah1607) [reply examples/kubernetes/storageclass-nfs.yaml:10]: Yes

**Review by andyzhangx — COMMENTED** [review]
overall LGTM, thanks for the contribution!

**Review by andyzhangx — APPROVED** [review]
/lgtm

File: examples/kubernetes/storageclass-nfs.yaml
Comment (andyzhangx) [reply examples/kubernetes/storageclass-nfs.yaml:10]: pls fix this later.

**Review by andyzhangx — APPROVED** [review]
/approve

**Pull request merged**
